state/protocol/protocol_state/mutator.go

```diff
@@ -269,3 +273,16 @@ func (m *stateMutator) transitionToEpochFallbackMode(results []*flow.ExecutionRe
 	}
 	return dbUpdates, nil
 }
+
+// epochFallbackTriggeredByIncorporatingCandidate checks whether incorporating the input block
+// would trigger epoch fallback mode (EFM) along the current fork. In particular, we trigger epoch
+// fallback mode when:
+//  1. The next epoch has not been committed as of B (EpochPhase < flow.EpochPhaseCommitted) AND
+//  2. B is the first incorporated block with view greater than or equal to the epoch commitment
+//     deadline for the current epoch
+func epochFallbackTriggeredByIncorporatingCandidate(candidateView, safetyThreshold, finalView uint64, phase flow.EpochPhase) bool {
+	if phase == flow.EpochPhaseCommitted { // Requirement 1
+		return false
+	}
+	return candidateView+safetyThreshold >= finalView // Requirement 2
+}
```

Comment:
Algorithmically it is correct I think, but I got stuck on it because there were a few details that I wasn't aware of. I think we can increase code clarity and documentation coverage notably if we changed the signature of the method just a bit.



```suggestion
// epochFallbackTriggeredByIncorporatingCandidate checks whether incorporating the input block B
// would trigger epoch fallback mode [EFM] along the current fork. We trigger epoch fallback mode
// when:
//  1. The next epoch has not been committed as of B (EpochPhase ≠ flow.EpochPhaseCommitted) AND
//  2. B is the first incorporated block with view greater than or equal to the epoch commitment
//     deadline for the current epoch
//
// In protocol terms, condition 1 means that an EpochCommit service event for the upcoming epoch has
// not yet been sealed as of block B. Formally, a service event S is considered sealed as of block B if:
//   - S was emitted during execution of some block A, s.t. A is an ancestor of B.
//   - The seal for block A was included in some block C, s.t C is an ancestor of B.
//
// For further details see `params.EpochCommitSafetyThreshold()`.
func epochFallbackTriggeredByIncorporatingCandidate(candidateView uint64, params protocol.GlobalParams, parentState *flow.RichProtocolStateEntry) bool {
	if parentState.EpochPhase() == flow.EpochPhaseCommitted { // Requirement 1
		return false
	}
	return candidateView+params.EpochCommitSafetyThreshold() >= parentState.CurrentEpochSetup.FinalView // Requirement 2
}
```

Here, I have used `params` and `parentState` instead of the derived values, because then I can easily refer to them in the documentation

Reply:
Done in https://github.com/onflow/flow-go/pull/5108/commits/6ba3c78bc0f66b21addf00b5fbf3d51066b59c29